AnalysisTools/plugins/PATObjectValueMapEmbedder.cc

```diff
@@ -62,22 +80,70 @@ template<typename T>
 void PATObjectValueMapEmbedder<T>::produce(edm::Event& iEvent,
                                            const edm::EventSetup& iSetup)
 {
-  edm::Handle<edm::View<T> > in;
   std::unique_ptr<std::vector<T> > out(new std::vector<T>);
-  edm::Handle<edm::ValueMap<float> > values;
-
-  iEvent.getByToken(srcToken, in);
-  iEvent.getByToken(valToken, values);
+  edm::Handle<edm::View<T> > in;
+  iEvent.getByToken(srcToken_, in);
 
-  for(size_t i = 0; i < in->size(); ++i)
+  //TODO: template, cleanup 
+  if (floatValTokens_.size() != floatLabels_.size())
     {
-      edm::Ptr<T> t = in->ptrAt(i);
+      throw cms::Exception("InvalidParams")
+        << "Must have equal number of values and labels";
+    }
+  if (boolValTokens_.size() != boolLabels_.size())
+    {
+      throw cms::Exception("InvalidParams")
+        << "Must have equal number of values and labels";
+    }
+  if (intValTokens_.size() != intLabels_.size())
+    {
+      throw cms::Exception("InvalidParams")
+        << "Must have equal number of values and labels";
+    }
+  std::vector<size_t> lengths = {floatLabels_.size(), boolLabels_.size(), intLabels_.size()};
 
-      out->push_back(*t);
-      out->back().addUserFloat(label, (*values)[t]);
+  for(size_t j = 0; j < in->size(); ++j)
+    {
+      edm::Ptr<T> t = in->ptrAt(j);
+      T obj = in->at(j);
+      for (size_t i = 0; i < (*std::max_element(lengths.begin(), lengths.end())); i++)
+        {
+          if(i < floatLabels_.size())
+            {
+              edm::Handle<edm::ValueMap<float> > floatVals;
+              iEvent.getByToken(floatValTokens_[i], floatVals);
```

Comment:
Why do we get the values from the event once for each object. I'd prefer to invert the loops so it just happens once per event.